File: Rpg_Game/Proto/lua/User_LoginProto.lua
--用户登入
User_LoginProto = { ProtoCode = 10000, id = 0, username = "" }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
User_LoginProto.__index = User_LoginProto;

function User_LoginProto.New()
    local self = { }; --初始化self
    setmetatable(self, User_LoginProto); --将self的元表设定为Class
    return self;
end


--发送协议
function User_LoginProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.id);
    ms:WriteUTF8String(proto.username);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function User_LoginProto.GetProto(buffer)

    local proto = User_LoginProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.id = ms:ReadInt();
    proto.username = ms:ReadUTF8String();

    ms:Dispose();
    return proto;
end

File: Rpg_Game/Proto/cshap/User_LoginProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2024-04-10 18:00:32
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 用户登入
/// </summary>
public struct User_LoginProto : IProto
{
    public ushort ProtoCode { get { return 10000; } }

    public int id; //
    public string username; //

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(id);
            ms.WriteUTF8String(username);
            return ms.ToArray();
        }
    }

    public static User_LoginProto GetProto(byte[] buffer)
    {
        User_LoginProto proto = new User_LoginProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.id = ms.ReadInt();
            proto.username = ms.ReadUTF8String();
        }
        return proto;
    }
}